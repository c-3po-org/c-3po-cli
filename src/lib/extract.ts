import "../declarations";
import * as babel from "@babel/core";
import * as fs from "fs";
import * as tmp from "tmp";
import babelPluginTtag from "babel-plugin-ttag";
import * as babelPresetReact from "@babel/preset-react";
import * as ttagTypes from "../types";
import { TransformFn, pathsWalk } from "./pathsWalk";

export async function extractAll(
    paths: string[],
    lang: string,
    progress: ttagTypes.Progress
): Promise<string> {
    const tmpFile = tmp.fileSync();
    let c3pOptions: ttagTypes.TtagOpts = { extract: { output: tmpFile.name } };
    if (lang !== "en") {
        c3pOptions.defaultLang = lang;
    }
    const babelOptions = {
        presets: [babelPresetReact],
        plugins: [[babelPluginTtag, c3pOptions]]
    };
    const transformFn: TransformFn = filepath => {
        try {
            babel.transformFileSync(filepath, babelOptions);
        } catch (err) {
            if (err.codeFrame) {
                console.error(err.codeFrame);
            } else {
                console.error(err);
            }
            progress.fail("Failed to extract translations");
            process.exit(1);
            return;
        }
    };
    await pathsWalk(paths, progress, transformFn);
    const result = fs.readFileSync(tmpFile.name).toString();
    tmpFile.removeCallback();
    return result;
}
