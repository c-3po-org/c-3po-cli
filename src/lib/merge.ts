import { PoData, Translations, Messages, Message, Comments } from "./parser";

/* merge two context maps together */
function mergeTranslations(
    leftContext: Translations,
    rightContext: Translations
): Translations {
    // Update messages known to the left side
    const merged: Translations = {};
    for (const contextKey of Object.keys(leftContext)) {
        if (!rightContext[contextKey]) {
            merged[contextKey] = leftContext[contextKey];
        } else {
            merged[contextKey] = mergeMessages(
                leftContext[contextKey],
                rightContext[contextKey]
            );
        }
    }
    // Append messages from the right side
    for (const contextKey of Object.keys(rightContext)) {
        if (!leftContext[contextKey]) {
            merged[contextKey] = rightContext[contextKey];
        }
    }
    return merged;
}
/* Merge two message maps together */
function mergeMessages(
    leftMessages: Messages,
    rightMessages: Messages
): Messages {
    const merged: Messages = {};
    // Update messages from left and merge with right
    for (const msgid of Object.keys(leftMessages)) {
        if (!rightMessages[msgid]) {
            merged[msgid] = leftMessages[msgid];
        } else {
            merged[msgid] = mergeMessage(
                leftMessages[msgid],
                rightMessages[msgid]
            );
        }
    }
    // Append messages from right
    for (const msgid of Object.keys(rightMessages)) {
        if (!leftMessages[msgid]) {
            merged[msgid] = rightMessages[msgid];
        }
    }
    return merged;
}

/* Merge two po(t)files together */
export function mergeMessage(
    leftMessage: Message,
    rightMessage: Message
): Message {
    if (leftMessage.msgid == "") {
        //gettext-parser library bug: header appears to be in message with id''
        return leftMessage;
    }
    const msgstr =
        leftMessage.msgstr.filter(s => !!s).length > 0
            ? leftMessage.msgstr
            : rightMessage.msgstr;
    return {
        msgid: leftMessage.msgid,
        comments: mergeComments(leftMessage.comments, rightMessage.comments),
        msgstr: msgstr,
        msgid_plural: leftMessage.msgid_plural,
        msgctxt: leftMessage.msgctxt
    };
}

/* Merge comments from two messages */
function mergeComments(
    leftComment: Comments | undefined,
    rightComment: Comments | undefined
): Comments | undefined {
    if (!leftComment) {
        return rightComment;
    }
    if (!rightComment) {
        return undefined;
    }
    if (!leftComment.reference) {
        return {
            reference: rightComment.reference
        };
    }
    if (!rightComment.reference) {
        return {
            reference: rightComment.reference
        };
    }
    const uniqueComments = new Set(
        leftComment.reference
            .split("\n")
            .concat(rightComment.reference.split("\n"))
    );
    const references = Array.from(uniqueComments.values()).sort();
    return {
        reference: references.join("\n")
    };
}

/* Merge two poData objects together */
export function mergePo(left: PoData, right: PoData): PoData {
    return {
        headers: left.headers,
        translations: mergeTranslations(left.translations, right.translations)
    };
}
