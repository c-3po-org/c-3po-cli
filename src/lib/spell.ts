import "../declarations";
import * as fs from "fs";
import * as os from "os";
import * as path from "path";
import fetch from "node-fetch";
import * as serialize from "serialize-javascript";

import * as Spellchecker from "hunspell-spellchecker";

const HOMEDIR = os.homedir();
const C3PODIR = ".ttag";
const DICTDIR = "dicts";

export interface Checker {
    check: (s: string) => boolean;
    suggest: (s: string) => string[];
}

/* Create necessary dirs */
function ensureDictDir() {
    if (!fs.existsSync(path.join(HOMEDIR, C3PODIR))) {
        fs.mkdirSync(path.join(HOMEDIR, C3PODIR));
    }
    if (!fs.existsSync(path.join(HOMEDIR, C3PODIR, DICTDIR))) {
        fs.mkdirSync(path.join(HOMEDIR, C3PODIR, DICTDIR));
    }
}

function getDictName(locale: string): string {
    return path.join(HOMEDIR, C3PODIR, DICTDIR, `${locale}.json`);
}

/* Safely serialize dict with regexp rules and save it */
function saveDict(dictName: string, dict: Object) {
    ensureDictDir();
    fs.writeFileSync(dictName, serialize(dict));
}

/* Load dict from file and eval it to create all objects */
function loadDict(dictName: string) {
    return eval(`(${fs.readFileSync(dictName).toString()})`);
}

/* Load dict file either from web or from disc and save it */
export async function getChecker(locale: string): Promise<Checker> {
    if (!LOCALEDICTMAP[locale]) {
        throw "Incorrect or unsupported locale";
    }
    const dictName = getDictName(locale);
    if (fs.existsSync(dictName)) {
        try {
            return new Spellchecker(loadDict(dictName));
        } catch (err) {
            console.error(err);
        }
    }
    const responseAff = await fetch(`${BASEPATH}${LOCALEDICTMAP[locale].aff}`);
    const aff = await responseAff.text();
    const responseDic = await fetch(`${BASEPATH}${LOCALEDICTMAP[locale].dic}`);
    const dic = await responseDic.text();
    const spellchecker = new Spellchecker();
    const dict = spellchecker.parse({ aff, dic });
    try {
        saveDict(dictName, dict);
    } catch (err) {
        console.warn(`Could not save dict to ${dictName}: ${err}`);
    }
    spellchecker.use(dict);
    return spellchecker;
}

const BASEPATH =
    "https://raw.githubusercontent.com/LibreOffice/dictionaries/master/";

const LOCALEDICTMAP: {
    [locale: string]: { [key in "aff" | "dic"]: string };
} = {
    uk: { aff: "uk_UA/uk_UA.aff", dic: "uk_UA/uk_UA.dic" },
    ru: { aff: "ru_RU/ru_RU.aff", dic: "ru_RU/ru_RU/dic" },
    en: { aff: "en/en_US.aff", dic: "en/en_US.dic" }
};
