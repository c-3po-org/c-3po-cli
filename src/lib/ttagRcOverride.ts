import { TtagRc } from "../types";
import * as fs from "fs";

type RC_TMPL = {
    extractor?: {
        ignoreFiles?: string;
        paths?: string;
    };
};

function readTtagRC(): TtagRc {
    const opts: any = {};
    try {
        const jsonRaw = fs.readFileSync(".ttagrc", "utf8");
        const parsedJSON: RC_TMPL = JSON.parse(jsonRaw);
        if ("extractor" in parsedJSON) {
            opts.extractor = parseExtractor(parsedJSON.extractor);
        }
    } catch (err) {
        console.warn(err);
    }
    return <TtagRc>opts;
}

function parseExtractor(extractorData: any) {
    const opts: any = {};
    if ("paths" in extractorData) {
        opts.paths = parseStringList(extractorData.paths);
    }
    if ("ignoreFiles" in extractorData) {
        opts.ignoreFiles = parseStringList(extractorData.ignoreFiles);
    }
    return opts;
}

function parseStringList(value: string): string[] {
    return value.split(",").map(x => x.trim());
}

export function parseTtagRcOpts(): TtagRc | undefined {
    const opts = readTtagRC();
    if (!Object.keys(opts)) return undefined;
    return opts;
}
