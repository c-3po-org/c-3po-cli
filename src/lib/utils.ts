const pluralNumRegex = /^nplurals ?= ?(\d);/;
export function getPluralFormsNumber(pluralFormsHeader: string): number {
    const match = pluralNumRegex.exec(pluralFormsHeader);
    if (match === null) {
        throw new Error(
            `Failed to parse plural-form header ${pluralFormsHeader}`
        );
    }
    let pluralFnCount = match[1];
    if (pluralFnCount[pluralFnCount.length - 1] === ";") {
        pluralFnCount = pluralFnCount.slice(0, -1);
    }
    return parseInt(pluralFnCount, 10);
}
