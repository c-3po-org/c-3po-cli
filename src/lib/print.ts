import chalk from "chalk";
import { Comments, Message, Headers } from "./parser";
import { generateHeader } from "gettext-parser/lib/shared";

/* Print formatted header */
export function printHeader(headers: Headers) {
    process.stdout.write(chalk.blue(generateHeader(headers)));
}

/* Print formatted comments if exists */
export function printComments(comments: Comments | undefined) {
    if (!comments) {
        return;
    }
    if (!comments.reference) {
        return;
    }
    for (const comment of comments.reference.split("\n")) {
        process.stdout.write(chalk.blue(`#: ${comment}\n`));
    }
    if (comments.flag === "fuzzy") {
        process.stdout.write(chalk.blue("#, fuzzy\n"));
    }
}

/* Print formatted context if exists */
export function printContext(ctxt: string) {
    if (ctxt != "") {
        process.stdout.write(`${chalk.yellow("msgctxt")} "${ctxt}"\n`);
    }
}

/* Print formatted msgid */
export function printMsgid(msgid: string) {
    process.stdout.write(`${chalk.yellow("msgid")} "${msgid}"\n`);
}

/* Print formatted msgstr */
export function printMsgstr(msgstr: string[]) {
    if (msgstr.length > 1) {
        for (let i = 0; i < msgstr.length; i++) {
            process.stdout.write(
                `${chalk.yellow(`msgstr[${i}]`)} "${msgstr[i]}"\n`
            );
        }
    } else {
        process.stdout.write(`${chalk.yellow("msgstr")} "${msgstr[0]}"\n`);
    }
}

/* Print formatted msgid plural*/
export function printMsgidPlural(msgid_plural: string | undefined) {
    if (!msgid_plural) {
        return;
    }
    process.stdout.write(
        `${chalk.yellow("msgid_plural:")} "${msgid_plural}"\n`
    );
}

/* Print full message */
export function printMsg(msg: Message) {
    printComments(msg.comments);
    printContext(msg.msgctxt || "");
    printMsgid(msg.msgid);
    printMsgidPlural(msg.msgid_plural);
    printMsgstr(msg.msgstr);
}
