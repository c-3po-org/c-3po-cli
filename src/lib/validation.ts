import { ExpressionStatement, TemplateLiteral } from "@babel/types";
import { ast2Str } from "./utils";
import tpl from "@babel/template";

interface FormatCheckResult {
    missing: string[];
    redundant: string[];
    valid: boolean;
}

const isoCodes =
    "http://docs.translatehouse.org/projects/localization-guide/en/latest/l10n/pluralforms.html";

export function langValidationMsg(language: string): string {
    return `Unknown lang code "${
        language
    }".\nSee all available lang codes here - ${isoCodes}`;
}

/* Parse template string with babel and return a Set of template identifiers and tagged expressions */
export function parseTemplateString(str: string): Set<string> {
    const templates: Set<string> = new Set();
    const expressionStmt = <ExpressionStatement>tpl.ast("`" + str + "`");
    // I cannot into types
    const expression = <TemplateLiteral>expressionStmt.expression;
    for (const node of expression.expressions) {
        if (
            node.type == "Identifier" ||
            node.type == "CallExpression" ||
            node.type == "TaggedTemplateExpression" ||
            node.type == "BinaryExpression" ||
            node.type == "MemberExpression"
        ) {
            templates.add(ast2Str(node));
        }
    }
    return templates;
}

/* Compare to sets of templates and return missing/redundant lists */
export function checkFormat(msgid: string, msgstr: string): FormatCheckResult {
    const result = <FormatCheckResult>{
        valid: true,
        missing: [],
        redundant: []
    };
    if (msgid.indexOf("${") == -1 && msgstr.indexOf("${") == -1) {
        return result;
    }
    let parsedId, parsedStr: Set<string>;
    try {
        parsedId = parseTemplateString(msgid);
        parsedStr = parseTemplateString(msgstr);
    } catch (e) {
        result.valid = false;
        return result;
    }
    for (const elem of parsedId) {
        if (!parsedStr.has(elem)) {
            result.missing.push(elem);
            result.valid = false;
        }
    }
    for (const elem of parsedStr) {
        if (!parsedId.has(elem)) {
            result.redundant.push(elem);
            result.valid = false;
        }
    }
    return result;
}
