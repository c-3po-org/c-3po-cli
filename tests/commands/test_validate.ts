import * as path from "path";
import { parseTemplateString, checkFormat } from "../../src/lib/validation";
import { execSync } from "child_process";

const poPath = path.resolve(__dirname, "../fixtures/validateTest/validate.po");

test("parse template string", () => {
    const identifiers = parseTemplateString("${x} and ${y}");
    expect(Array.from(identifiers.values()).join("|")).toBe("x|y");
    const callexpressions = parseTemplateString("${x()} and ${y()}");
    expect(Array.from(callexpressions.values()).join("|")).toBe("x()|y()");
    const taggedexpression = parseTemplateString("${x + z} and ${y()*5}");
    expect(Array.from(taggedexpression.values()).join("|")).toBe(
        "x + z|y() * 5"
    );
});

test("invalid format checks", () => {
    const result = checkFormat("${x}", "${y}");
    expect(result.valid).toBe(false);
    expect(result.missing.join("")).toBe("x");
    expect(result.redundant.join("")).toBe("y");
});

test("command integrity test", () => {
    try {
        execSync(`ts-node src/index.ts validate ${poPath}`);
    } catch (err) {
        expect(err.status).toBe(1);
    }
});
