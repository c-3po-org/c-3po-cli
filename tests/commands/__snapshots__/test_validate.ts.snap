// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`command integrity test 1`] = `
"[34m#: tests/fixtures/testset.js:2[39m
[34m[39m[33mmsgid[39m \\"test_broken \${foo.bar}\\"
[33mmsgstr[39m \\"[4m[41mwrong_interpolation \${bar.foo}[49m[24m <--- [32mmissing foo.bar but redundant bar.foo[39m;\\"


[33mmsgctxt[39m \\"xxx\\"
[33mmsgid[39m \\"test_broken \${y}_fuzzy_no_ref\${x(13)}\\"
[33mmsgstr[39m \\"[4m[41mtail\${y} \${x(12) + 12} \${z}[49m[24m <--- [32mmissing x(13) but redundant x(12) + 12 and z[39m;\\"


"
`;
